go-packages/sim2/car.go

```diff
@@ -1,163 +1,160 @@
+    c.path.riderAddress = address
+    c.requestState = Success
+  } else {
+    log.Println("Car ",c.id," Accept Request failed")
+    c.requestState = Fail
+  }
+}
 
-  // Slope of edge and perpendecular section
-  mEdge := (y2 - y1) / (x2 - x1)
-  mPerp := float64(-1) / mEdge
+func (c *Car) getLocations() (pickup Location, dropOff Location) {
+  from, to := c.ethApi.GetLocations(c.path.riderAddress)
+  fmt.Println("Car",c.id," locations ",from," ", to)
+  x, y := splitCSV(from)
+  pickup = c.world.closestEdgeAndCoord(Coords{x,y})
+  x, y = splitCSV(to)
+  dropOff = c.world.closestEdgeAndCoord(Coords{x,y})
+  return
+}
 
-  //fmt.Print("<checkIntersect>")
-  //fmt.Print(" query: {", xQuery, ",", yQuery, "}")
-  //fmt.Print(", Edge{", x1, ",", y1, "}->{", x2, ",", y2, "}")
-  //fmt.Print(", mEdge: ", mEdge, ", mPerp: ", mPerp)
+func (c *Car) getShortestPathToEdge(edge Edge) (edges []Edge, dist float64) {
+  return c.world.ShortestPath(c.path.currentEdge.End.ID, edge.Start.ID)
+}
 
-  inPerp := false  // By default, not in perpendicular region
-  isVert := math.IsInf(mPerp, 0)  // Is the perpendicular section vertical?
 
-  // Determine if the query point lies in region perpendicular to the edge segment
-  if (isVert) {  // Regions determined by x values
-    if ((x1 < x2 && x1 < xQuery && xQuery < x2) || (x2 < x1 && xQuery < x1 && x2 < xQuery)) {
-      inPerp = true
-    }
-  } else {  // Regions determined by y values
-    // Relative y-coordinates of perpendicular lines at bounds of edge segment at x-value of query
-    y1Rel := mPerp * (xQuery - x1) + y1
-    y2Rel := mPerp * (xQuery - x2) + y2
-    if ((y1Rel < y2Rel && y1Rel < yQuery && yQuery < y2Rel) || (y2Rel < y1Rel && yQuery < y1Rel && y2Rel < yQuery)) {
-      inPerp = true
-    }
+func (c *Car) drive() {
+  if c.path.currentState == DrivingAtRandom {
+	c.checkRequestState()
   }
-
-  //fmt.Print(", inPerp: ", inPerp, ", isVert: ", isVert)
-
-  if (!inPerp) {  // Not in perpendicular segment or vertical; identify closest
-    distance1 := query.Distance(e.Start.Pos)
-    distance2 := query.Distance(e.End.Pos)
-    //fmt.Println(", dist1: ", distance1, ", dist2: ", distance2)
-    if (distance1 < distance2) {
-      intersect = e.Start.Pos
-      distance = distance1
-    } else {
-      intersect = e.End.Pos
-      distance = distance2
-    }
-  } else {  // In perpendicular region; find intersection on Edge
-    // Check for straight lines to ease calculations
-    if (x1 == x2) {
-      intersect.X = math.Round(x1)
-      intersect.Y = math.Round(yQuery)
-
-    } else if (y1 == y2) {
-      intersect.X = math.Round(xQuery)
-      intersect.Y = math.Round(y1)
-    } else {
-      intX := (mEdge * x1 - mPerp * xQuery + yQuery - y1) / (mEdge - mPerp)
-      intY := mPerp * (intersect.X - xQuery) + yQuery
-      intersect.X = math.Round(intX)
-      intersect.Y = math.Round(intY)
+  if c.path.currentPos.Equals(c.path.currentEdge.End.Pos) {  // Already at edge end, so change edge
```

Comment:
I don't think this current logic permits the car to actually stop at the designated drop-off location; only the nearest edge endpoint. I also think that we should probably invert this logic: have the car evaluate its path planning state by reading the map, not the other way around. Something to the tune of:
```go
switch c.path.currentState {
case DrivingAtRandom:
  // if at end of route, select new random destination
case ToPickUp:
  // check current position
    // if not on the edge, walk all edges until at the target edge
    // if on correct edge and not at target coords, start looking for pickup coords on the edge
    // if at the target coords, enter 'stopped' state until user indicates via UI 'ready to go' -> enter ToDropOff state
case ToDropOff:
  // check current position
    // if not on the edge, walk all edges until at the target edge
    // if on correct edge and not at target coords, start looking for pickup coords on the edge
    // if at the target coords, enter 'stopped' state until user indicates via UI 'done with ride' -> enter DrivingAtRandom state
case Stopped:
  // await some kind of external input, probably from a channel feeding information from the web UI
}
```

Then after the car has decided where it wants to move (behavior), it then plans the simulated motion (actuation). That will involve calculating velocity and displacement along edges, preventing collisions, interacting with road rules.

My objective is to separate these two evaluative behaviors as much as possible. Does the rationale make sense?